text/0000-serializable-types.md

```diff
@@ -0,0 +1,122 @@
+- Feature Name: `serializable-types`
+- Start Date: 2020-11-06
+- RFC PR: [Unity-Technologies/com.unity.multiplayer.rfcs#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/pull/0000)
+- Issue: [Unity-Technologies/com.unity.multiplayer#0000](https://github.com/Unity-Technologies/com.unity.multiplayer/issues/0000)
+
+# Summary
+[summary]: #summary
+
+_RFC draft for **Serializable Types** (to be updated and properly proposed)_
+
+# Motivation
+[motivation]: #motivation
+
+Why are we doing this? What use cases does it support? What is the expected outcome?
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Serialization
+
+Multiplayer framework has built-in serialization support for C# and Unity primitive types out-of-the-box, also with ability to further extend network serialization for user defined types implementing `INetworkSerializable` interface.
+
+### C# Primitives
+
+`bool`, `char`, `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `string` types will be serialized by built-in serialization code.
+
+### Unity Primitives
+
+`Color`, `Color32`, `Vector2`, `Vector3`, `Vector4`, `Quaternion`, `Ray`, `Ray2D` types will be serialized by built-in serialization code.
+
+### Enum Types
+
+A user-defined enum type will be serialized by built-in serialization code (with underlying integer type).
+
+### Static Arrays
+
+Static arrays like `int[]` will be serialized by built-in serialization code if their underlying type is either one of serialization supported types (e.g. `Vector3`) or if they implement `INetworkSerializable` interface.
+
+### Generic Collections
+
+Generic collections such as `IEnumerable<T>` and `IEnumerable<KeyValuePair<K, V>>` will be serialized by built-in serialization code if their underlying type is either one of serialization supported types (e.g. `Vector3`) or if they implement `INetworkSerializable` interface.
+
+### INetworkSerializable
+
+Complex user-defined types that implements `INetworkSerializable` interface will be serialized by user provided serialization code:
```

Comment:
I dont know if this document is ready for discussion.

But i think there are two ways of doing it, having to implement an interface or registering serialization at some point. Currently [MLAPI does both](https://mlapi.network/wiki/custom-serialization/), which is confusing.

I think one important difference between the two is that when registering a custom serializer-per-type you have a lot more control. You can serialize types in assemblies you dont own, e.g. i buy a DLL from the asset store and i cannot access the code to make it implement INetworkSerializable. Or Unity types, i can register a serializer for a RigidBody and sync these with my custom method. When registering it is possible to add this functionality to existing classes, with interfaces it is not.

Another case is overwriting default behaviour. I might want to override how Unity Multiplayer sends `float`, because i know that in my game the world is much smaller, so i can get away with sending a `short` instead.

Reply:
you could always implement `MyExternalTypeSerializable : MyExternalType, INetworkSerializable` or simply put `MyExternalType` under another `INetworkSerializable` type and serialize as a part of it.

_(this document is not ready, just created to extract serialization topic out of RPC)_